src/DDMinimizer.cpp

```diff
@@ -0,0 +1,426 @@
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateLeft(layout, stairsCh);
+    } else if (prioXl == 0 && stairsXl > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateRight(layout, stairsXl);
+    } else if (prioCh > 0 || prioXl > 0 || (prioCh == 0 && prioXl == 0)) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+  } else if ((xCIndex != -1) && (cXIndex == -1 || cXIndex > xCIndex)) {
+
+    if ((prioCl == 0 && DDMinimizer::isFullLadder(cLIndex)) ||
+        (prioXh == 0 && DDMinimizer::isFullLadder(xHIndex))) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+
+    else if (prioCl == 0 && stairsCl > 0) {
+      layout = DDMinimizer::rotateRight(layout, stairsCl);
+    } else if (prioXh == 0 && stairsXh > 0) {
+      layout = DDMinimizer::rotateLeft(layout, stairsXh);
+    }
+  } else if ((DDMinimizer::isFullLadder(xHIndex) ||
+              DDMinimizer::isFullLadder(cLIndex))) {
+    layout = DDMinimizer::reverseLayout(layout);
+  } else {
+    // in case no full pattern was identified, call fallback function to
+    // determine ordering based on singular controlled operations
+    return DDMinimizer::createControlBasedPermutation(qc);
+  }
+
+  // transform layout into permutation
+  qc::Permutation perm;
+  std::vector<Qubit> layer(bits);
+  for (qc::Qubit i = 0; i < bits; i++) {
+    layer[i] = i;
+  }
+  for (qc::Qubit i = 0; i < bits; i++) {
+    perm[layer[i]] = layout[i];
+  }
+  return perm;
+}
+
+void DDMinimizer::initializeDataStructure(
+    std::size_t bits, std::map<std::pair<Qubit, Qubit>, int>& xCMap,
+    std::map<std::pair<Qubit, Qubit>, int>& cXMap,
+    std::vector<std::map<std::pair<Qubit, Qubit>, int>>& cLMap,
+    std::vector<std::map<std::pair<Qubit, Qubit>, int>>& cHMap,
+    std::vector<std::map<std::pair<Qubit, Qubit>, int>>& xLMap,
+    std::vector<std::map<std::pair<Qubit, Qubit>, int>>& xHMap) {
+  const std::size_t max = bits - 1;
+  // create x-c ladder
+  for (size_t i = 0; i < max; i++) {
+    xCMap.insert({{i + 1, i}, -1});
+    cXMap.insert({{i, i + 1}, -1});
+  }
+
+  // create c-l and x-l ladder
+  for (size_t i = 0; i < max; i++) {
+    for (size_t j = 0; j < bits; j++) {
+      if (i < j) {
+        xLMap[i].insert({{j, i}, -1});
+        cLMap[i].insert({{i, j}, -1});
+      }
+    }
+  }
+
+  // create c-h and x-h ladder
+  for (size_t i = max; i > 0; i--) {
+    for (size_t j = 0; j < bits; j++) {
+      if (i > j) {
+        xHMap[bits - i - 1].insert({{j, i}, -1});
+        cHMap[bits - i - 1].insert({{i, j}, -1});
+      }
+    }
+  }
+}
+
+int DDMinimizer::findMaxIndex(
+    const std::map<std::pair<Qubit, Qubit>, int>& map) {
+  int maxIndex = -1; // Initialize to the smallest possible value
+
+  for (const auto& entry : map) {
+    if (entry.second == -1) {
+      maxIndex = -1;
+      break;
+    }
+    if (entry.second > maxIndex) {
+      maxIndex = entry.second;
+    }
+  }
+
```

Comment:
From the name of the function and the code, I expect that it returns the maximum `value` of all `key`, `value` pairs (even though `findMaxIndex` might be missleading since `Index` would suggest key not value).

I do not understand why the `maxIndex` is reset to -1 if the `value` of -1 is encountered. Is there a reason for that? If yes, I think then the output of the function depends on the order the map returns the elements which can be arbitrary and the return value should be independent from the order. However, maybe I am overseeing here something. Then let me know.